File: fotoapparat/src/main/java/io/fotoapparat/parameter/selector/PreviewFpsRangeSelectors.java
package io.fotoapparat.parameter.selector;

import java.util.Collection;
import java.util.Collections;
import java.util.Comparator;

import io.fotoapparat.parameter.range.Range;
import io.fotoapparat.util.CompareFpsRangeByBounds;
import io.fotoapparat.util.ExactFpsRangePredicate;
import io.fotoapparat.util.InBoundsFpsRangePredicate;

import static io.fotoapparat.parameter.selector.Selectors.filtered;
import static io.fotoapparat.parameter.selector.Selectors.firstAvailable;

/**
 * Selector functions for preview FPS ranges.
 */
public class PreviewFpsRangeSelectors {

    private static final Comparator<Range<Integer>> COMPARATOR_BY_BOUNDS = new CompareFpsRangeByBounds();

    /**
     * For conversion fps to FPS range bound
     */
    private static final int FPS_RANGE_BOUNDS_SCALE = 1000;

    /**
     * @param fps the specified FPS
     * @return {@link SelectorFunction} which selects FPS range that contains only the specified FPS.
     */
    public static SelectorFunction<Collection<Range<Integer>>, Range<Integer>> fromExactFps(int fps) {
        return filtered(rangeWithHighestFps(),
                new ExactFpsRangePredicate(fps * FPS_RANGE_BOUNDS_SCALE));
    }

    /**
     * @param fps the specified FPS
     * @return {@link SelectorFunction} which selects FPS range that contains the specified FPS.
     */
    public static SelectorFunction<Collection<Range<Integer>>, Range<Integer>> nearestToExactFps(int fps) {
        return firstAvailable(fromExactFps(fps),
                filtered(rangeWithHighestFps(), new InBoundsFpsRangePredicate(fps * FPS_RANGE_BOUNDS_SCALE)));
    }

    /**
     * @return {@link SelectorFunction} which selects FPS range with max FPS.
     */
    public static SelectorFunction<Collection<Range<Integer>>, Range<Integer>> rangeWithHighestFps() {
        return new SelectorFunction<Collection<Range<Integer>>, Range<Integer>>() {
            @Override
            public Range<Integer> select(Collection<Range<Integer>> items) {
                if (items.isEmpty()) {
                    return null;
                }

                return Collections.max(items, COMPARATOR_BY_BOUNDS);
            }
        };
    }

    /**
     * @return {@link SelectorFunction} which selects FPS range with min FPS.
     */
    public static SelectorFunction<Collection<Range<Integer>>, Range<Integer>> rangeWithLowestFps() {
        return new SelectorFunction<Collection<Range<Integer>>, Range<Integer>>() {
            @Override
            public Range<Integer> select(Collection<Range<Integer>> items) {
                if (items.isEmpty()) {
                    return null;
                }

                return Collections.min(items, COMPARATOR_BY_BOUNDS);
            }
        };
    }

}

File: fotoapparat/src/main/java/io/fotoapparat/selector/PreviewFpsRangeSelectors.kt
package io.fotoapparat.selector

import io.fotoapparat.parameter.FpsRange
import io.fotoapparat.util.CompareFpsRangeByBounds

typealias FpsRangeSelector = Iterable<FpsRange>.() -> FpsRange?

/**
 * @param fps The specified FPS
 * @return Selector function which selects FPS range that contains the specified FPS.
 * Prefers fixed rates over non fixed ones.
 */
fun containsFps(fps: Float): FpsRangeSelector = firstAvailable(
        exactFixedFps(fps),
        filtered(
                selector = highestNonFixedFps(),
                predicate = { range -> fps.toFpsIntRepresentation() in range }
        )
)

/**
 * @param fps The specified FPS
 * @return Selector function which selects FPS range that contains only the specified FPS.
 */
fun exactFixedFps(fps: Float): FpsRangeSelector = filtered(
        selector = highestFixedFps(),
        predicate = { it.min == fps.toFpsIntRepresentation() }
)

/**
 * @return Selector function which selects FPS range with max FPS.
 * Prefers non fixed rates over fixed ones.
 */
fun highestFps(): FpsRangeSelector = firstAvailable(
        highestNonFixedFps(),
        highestFixedFps()
)

/**
 * @return Selector function which selects FPS range with max FPS and non fixed rate.
 */
fun highestNonFixedFps(): FpsRangeSelector = filtered(
        selector = highestRangeFps(),
        predicate = { !it.isFixed }
)

/**
 * @return Selector function which selects FPS range with max FPS and fixed rate.
 */
fun highestFixedFps(): FpsRangeSelector = filtered(
        selector = highestRangeFps(),
        predicate = { it.isFixed }
)

/**
 * @return Selector function which selects FPS range with min FPS.
 * Prefers non fixed rates over fixed ones.
 */
fun lowestFps(): FpsRangeSelector = firstAvailable(
        lowestNonFixedFps(),
        lowestFixedFps()
)

/**
 * @return Selector function which selects FPS range with min FPS and non fixed rate.
 */
fun lowestNonFixedFps(): FpsRangeSelector = filtered(
        selector = lowestRangeFps(),
        predicate = { !it.isFixed }
)

/**
 * @return Selector function which selects FPS range with min FPS and fixed rate.
 */
fun lowestFixedFps(): FpsRangeSelector = filtered(
        selector = lowestRangeFps(),
        predicate = { it.isFixed }
)

private fun highestRangeFps(): FpsRangeSelector = { maxWith(CompareFpsRangeByBounds) }

private fun lowestRangeFps(): FpsRangeSelector = { minWith(CompareFpsRangeByBounds) }

private fun Float.toFpsIntRepresentation() = (this * 1000).toInt()